drivers/filesystems/fastfat/misc.c

```diff
@@ -336,19 +336,64 @@ static
 VOID
 NTAPI
 VfatDoRequest(
-    PVOID IrpContext)
+    PVOID Context)
 {
+    PVFAT_IRP_CONTEXT IrpContext = Context;
+    PDEVICE_EXTENSION DeviceExt;
+    KIRQL OldIrql;
+
     InterlockedDecrement(&QueueCount);
-    DPRINT("VfatDoRequest(IrpContext %p), MajorFunction %x, %d\n",
-           IrpContext, ((PVFAT_IRP_CONTEXT)IrpContext)->MajorFunction, QueueCount);
-    VfatDispatchRequest((PVFAT_IRP_CONTEXT)IrpContext);
+
+    if (IrpContext->Stack->FileObject != NULL)
+    {
+        DeviceExt = IrpContext->Stack->DeviceObject->DeviceExtension;
+        ObReferenceObject(DeviceExt->VolumeDevice);
+    }
+
+    do
+    {
+        DPRINT("VfatDoRequest(IrpContext %p), MajorFunction %x, %d\n",
+               IrpContext, IrpContext->MajorFunction, QueueCount);
+        VfatDispatchRequest(IrpContext);
+        IrpContext = NULL;
+
+        /* Now process any overflow items */
+        if (DeviceExt != NULL)
```

Comment:
Would it make sense to also check OverflowQueueCount here outside the lock?
Theoretically this would prevent grabbing the spinlock when there is no work to be done.

Reply:
We do need to do work in either case, since `PostedRequestCount` is also protected by the lock.
I don't think it's a safe shortcut to take, since the queuing thread must also have a guarantee about the fact that there's a worker thread active that will handle its overflow request. So if we determine that `OverflowQueueCount` is 0, we must reset `PostedRequestCount` atomically with that check, or we risk letting an item sit in the queue without a thread to pick it up.